streams/src/main/java/org/apache/kafka/streams/processor/internals/namedtopology/KafkaStreamsNamedTopologyWrapper.java

```diff
@@ -118,40 +133,90 @@ public NamedTopologyBuilder newNamedTopologyBuilder(final String topologyName) {
         } else if (getTopologyByName(newTopology.name()).isPresent()) {
             throw new IllegalArgumentException("Unable to add the new NamedTopology " + newTopology.name() +
                                                    " as another of the same name already exists");
         }
-        topologyMetadata.registerAndBuildNewTopology(newTopology.internalTopologyBuilder());
+        return new AddNamedTopologyResult(
+            topologyMetadata.registerAndBuildNewTopology(newTopology.internalTopologyBuilder())
+        );
     }
 
     /**
      * Remove an existing NamedTopology from a running Kafka Streams app. If multiple instances of the application are
-     * running, you should inform all of them by calling {@link #removeNamedTopology(String)} on each client to ensure
+     * running, you should inform all of them by calling {@code #removeNamedTopology(String)} on each client to ensure
      * it stops processing the old topology.
      *
+     * @param topologyToRemove          name of the topology to be removed
+     * @param resetOffsets              whether to reset the committed offsets for any source topics
+     *
      * @throws IllegalArgumentException if this topology name cannot be found
      * @throws IllegalStateException    if streams has not been started or has already shut down
      * @throws TopologyException        if this topology subscribes to any input topics or pattern already in use
      */
-    public void removeNamedTopology(final String topologyToRemove) {
+    public RemoveNamedTopologyResult removeNamedTopology(final String topologyToRemove, final boolean resetOffsets) {
+        log.debug("Removing topology: {}", topologyToRemove);
         if (!isRunningOrRebalancing()) {
             throw new IllegalStateException("Cannot remove a NamedTopology while the state is " + super.state);
         } else if (!getTopologyByName(topologyToRemove).isPresent()) {
             throw new IllegalArgumentException("Unable to locate for removal a NamedTopology called " + topologyToRemove);
         }
+        final Set<TopicPartition> partitionsToReset = metadataForLocalThreads()
+            .stream()
+            .flatMap(t -> {
+                final HashSet<TaskMetadata> tasks = new HashSet<>();
+                tasks.addAll(t.activeTasks());
+                tasks.addAll(t.standbyTasks());
+                return tasks.stream();
+            })
+            .flatMap(t -> t.topicPartitions().stream())
+            .filter(t -> topologyMetadata.sourceTopologies(topologyToRemove).contains(t.topic()))
+            .collect(Collectors.toSet());
+
+
+
+        final KafkaFuture<Void> removeTopologyFuture = topologyMetadata.unregisterTopology(topologyToRemove);
+
+        if (resetOffsets) {
+            log.info("partitions to reset: {}", partitionsToReset);
+            if (!partitionsToReset.isEmpty()) {
+                try {
+                    removeTopologyFuture.get();
```

Comment:
We shouldn't call `get()` on the futures, that's up to the user to do. Technically speaking the point of this isn't to make the APIs blocking, but to enable users to block on the completion of them -- ie, we want to return the `RemoveNamedTopologyResult` ASAP and allow the user to block on it, if they want

Reply:
unfortunately if we want to reset the offsets we need to make sure we are not subscribed. otherwise we get : 

```java.util.concurrent.ExecutionException: org.apache.kafka.common.errors.GroupSubscribedToTopicException: Deleting offsets of a topic is forbidden while the consumer group is actively subscribed to it.
```

Reply:
Hm, I see. Ok that's potentially going to be a bit of an issue with multi-node clusters...let's chat about this tomorrow, I have some thoughts on a few different approaches we might want to consider here.

That said, we still should not call `get()` here, can we just move the offset reset into `TopologyMetadata` and do it after the last thread has processed the topology removal? I know it's similar but I really think we should make sure to return from `removeNamedTopology()` ASAP so that it's async and the caller can in theory do some other stuff if they wanted to before blocking on the result

Reply:
Yeah I hadn't thought of multi node clusters, that will be a problem. calling get here isn't going to be enough even if we wanted to do that.

We can maybe do it so that it will call it for each cluster after it has unsubscribed and expect it to fail for all but the last one? we can talk tomorrow

Reply:
how about something like this:
`        if (resetOffsets) {
            KafkaFutureImpl<Void> future = new KafkaFutureImpl<>();
            log.info("partitions to reset: {}", partitionsToReset);
            if (!partitionsToReset.isEmpty()) {
                removeTopologyFuture.whenComplete( (v, e) -> {
                    DeleteConsumerGroupOffsetsResult deleteOffsetsResult = null;
                    while (deleteOffsetsResult == null) {
                        try {
                            deleteOffsetsResult = adminClient.deleteConsumerGroupOffsets(
                                applicationConfigs.getString(StreamsConfig.APPLICATION_ID_CONFIG), partitionsToReset);
                            deleteOffsetsResult.all().get();
                        } catch (InterruptedException ex) {
                            ex.printStackTrace();
                            break;
                        } catch (ExecutionException ex) {
                            ex.printStackTrace();
                        }
                        try {
                            Thread.sleep(100);
                        } catch (InterruptedException ex) {
                            ex.printStackTrace();
                        }
                    }
                    future.complete(null);
                    });
                return new RemoveNamedTopologyResult(removeTopologyFuture, future);
            }
        }`

Reply:
synced off line